### ISSUE TITLE
Flushing specific cache with redis flush all caches

### ISSUE THREAD

**Issue by carloricci84** [magento/magento2#26020]
### Summary 

When we try to flush a specific cache type all redis database is flushed

### Preconditions (*)

1. Use Magento 2.3.3(2.4-develop) installed (Previous version have the same problem)
2. Use cm_backend_cache_redis
```
'cache' => [
        'frontend' => [
            'default' => [
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => '127.0.0.1',
                    'database' => '0',
                    'port' => '6379'
                ],
            ],
            'page_cache' => [
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => '127.0.0.1',
                    'port' => '6379',
                    'database' => '1',
                    'compress_data' => '0'
                ]
            ]
        ]
    ],
    'cache_types' => [
        'config' => 1,
        'layout' => 1,
        'block_html' => 1,
        'collections' => 1,
        'reflection' => 1,
        'db_ddl' => 1,
        'eav' => 1,
        'customer_notification' => 1,
        'config_integration' => 1,
        'config_integration_api' => 1,
        'target_rule' => 1,
        'full_page' => 1,
        'translate' => 1,
        'config_webservice' => 1,
        'compiled_config' => 1
    ],
```

### Steps to reproduce (*)

1. Flush all caches
2. Load the homepage
3. Run **redis-cli --bigkeys** or any method that will give you the number of keys in your redis cache
4. Run **php bin/magento cache:flush config** or run any scrypt that should flush one specific cache type.
5. Run **redis-cli --bigkeys** or any method that will give you the number of keys in your redis cache

### Expected result (*)

1. Only config keys are flushed

### Actual result (*)

1. All keys are flushed
<img width="971" alt="Screen Shot 2019-12-12 at 1 39 30 PM" src="https://user-images.githubusercontent.com/6392435/70741726-f1541a80-1ce9-11ea-82d8-078d96bdd039.png">

### More information
Looking at the code it is normal that all cache are flushed, it should be changed to only flush the required cache

In Magento\Backend\Console\Command\CacheFlushCommand The function flush of the cache manager is called
```
protected function performAction(array $cacheTypes)
    {
        $this->eventManager->dispatch('adminhtml_cache_flush_all');
        $this->cacheManager->flush($cacheTypes);
    }
```

In \Magento\Framework\App\Cache\Manager $backend->clean() is called, $backend here correspond to the class Cm_Cache_Backend_Redis. Note that there is no argument passed to the call.

```
public function flush(array $types)
    {
        $flushedBackend = [];
        foreach ($types as $type) {
            $backend = $this->pool->get($type)->getBackend();
            if (in_array($backend, $flushedBackend, true)) { // it was already flushed from another frontend
                continue;
            }

            $backend->clean();
            $flushedBackend[] = $backend;
        }
    }
```
In Cm_Cache_Backend_Redis (/vendor/colinmollenhour/cache-backend-redis/Cm/Cache/Backend/Redis.php) if no argument is passed to the function clean, it flush all redis DB

```
public function clean($mode = Zend_Cache::CLEANING_MODE_ALL, $tags = array())
    {
        if( $tags && ! is_array($tags)) {
            $tags = array($tags);
        }

        try {
            if ($mode == Zend_Cache::CLEANING_MODE_ALL) {
                return $this->_redis->flushDb();
            }
            if ($mode == Zend_Cache::CLEANING_MODE_OLD) {
                $this->_collectGarbage();
                return TRUE;
            }
            if ( ! count($tags)) {
                return TRUE;
            }
            switch ($mode)
            {
                case Zend_Cache::CLEANING_MODE_MATCHING_TAG:

                    $this->_removeByMatchingTags($tags);
                    break;

                case Zend_Cache::CLEANING_MODE_NOT_MATCHING_TAG:

                    $this->_removeByNotMatchingTags($tags);
                    break;

                case Zend_Cache::CLEANING_MODE_MATCHING_ANY_TAG:

                    $this->_removeByMatchingAnyTags($tags);
                    break;

                default:
                    Zend_Cache::throwException('Invalid mode for clean() method: '.$mode);
            }
        } catch (CredisException $e) {
            Zend_Cache::throwException('Error cleaning cache by mode '.$mode.': '.$e->getMessage(), $e);
        }
        return TRUE;
    }
```
### Proposed solution
Params should be passed to $backend->clean() if the backend class is Cm_Cache_Backend_Redis to prevent all cache from beeing flushed.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Delta 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-29935` were created

**Issue Available**: @engcom-Delta, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by matiashidalgo** [comment]
Not sure if my assumption is right but after reviewing this issue and debugging for a while I realize that the way this works is actually what is expected to happen.

Explanation: if we take into account the console description for cache:flush it says "Flushes cache storage used by cache type(s)" then if we consider a local setup where redis is the cache storage used by config, block_html, layout, collections, etc. then it is expected to wipe out all of them if we call "cache:flush config" because this console action is not meant to clean config cache type, it is meant to flush it's cache storage, which is redis db where it is stored many others cache types.

Extra notes: This same behaviour we can reproduce from admin panel where we can choose different cache types for "refresh" which indeed means clean, but we can only use one button for Flush Cache Storage which finally flushes all the backend cache storages.

Solution: instead of use cache:flush config use cache:clean config, this another console command is meant to be used for clean an specific cache type as it's own description says "Cleans cache type(s)"

**Comment by engcom-Golf** [comment]
Hello, @carloricci84 
According to documentation cache:flush command flushing a cache type purges the cache storage, which might affect other processes applications that are using the same storage.
It means if different cache types use the same database in Redis they will be flushed as well because it flushes the whole storage.
To avoid flushing all cache types use cache:clean [type] command to clean specific cache types by tags.
Please see https://devdocs.magento.com/guides/v2.4/config-guide/cli/config-cli-subcommands-cache.html#config-cli-subcommands-cache-clean
I will close this ticket as not an issue.
Thank you for your contribution.

**Comment by magento-engcom-team** [comment]
Hi @carloricci84, @engcom-Golf. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-29935` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
